### PR TITLE
Split SagaPersistenceBehavior into LoadSagaBehavior and PersistSagaBehavior, inlined InvokeSagaNotFound

### PR DESCRIPTION
**Pull request by danielmarbach**
When I was spelunking in the code base for #3241 I saw that `SagaPersistenceBehavior` is a small but [subtle issue](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Sagas/SagaPersistenceBehavior.cs#L25). I dediced to fix it and split `SagaPersistenceBehavior` into two behavior. One loads the saga information and the other one persists the saga information. When you inline the `InvokeSagaNotFoundBehavior` into the `PersistSagaBehavior` you no longer need `SagaInvocationResult` because all the required info is available on `ActiveSagaInstance`

@Particular/nservicebus please review

### CODE REVIEW

File: src/NServiceBus.Core/Sagas/LoadSagaBehavior.cs [hunk 1]
```diff
@@ -273,12 +219,18 @@ IContainSagaData CreateNewSagaEntity(SagaMetadata metadata, IInvokeHandlerContex
...
-        ICancelDeferredMessages timeoutCancellation;
 
-        static ILog logger = LogManager.GetLogger<SagaPersistenceBehavior>();
+        static ILog logger = LogManager.GetLogger<LoadSagaBehavior>();
+
+        public class Registration : RegisterStep
+        {
+            public Registration() : base("LoadSagaBehavior", typeof(LoadSagaBehavior), "Loads a saga if any.")
+            {
+                InsertBeforeIfExists(WellKnownStep.InvokeSaga);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/LoadSagaBehavior.cs]: :crying_cat_face: We need to figure out a way to register the behaviors in a given order instead (not this pull)

File: src/NServiceBus.Core/Sagas/Sagas.cs [hunk 2]
```diff
@@ -33,13 +32,13 @@ internal Sagas()
...
-        ///     See <see cref="Feature.Setup" />.
+        /// See <see cref="Feature.Setup" />.
         /// </summary>
         protected internal override void Setup(FeatureConfigurationContext context)
         {
             // Register the Saga related behaviors for incoming messages
-            context.Pipeline.Register(WellKnownStep.InvokeSaga, typeof(SagaPersistenceBehavior), "Invokes the saga logic");
-            context.Pipeline.Register("InvokeSagaNotFound", typeof(InvokeSagaNotFoundBehavior), "Invokes saga not found logic");
+            context.Pipeline.Register<PersistSagaBehavior.Registration>();
+            context.Pipeline.Register<LoadSagaBehavior.Registration>();
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/Sagas.cs]: @SzymonPobiega wasn't there a syntax to enforce order here? (not a blocker for this pull)

File: src/NServiceBus.Core/Sagas/LoadSagaBehavior.cs [hunk 3]
```diff
@@ -57,63 +52,14 @@ public SagaPersistenceBehavior(ISagaPersister persister, ICancelDeferredMessages
...
-                    }
                 }
             }
             else
             {
-                context.Extensions.Get<SagaInvocationResult>().SagaFound();
                 sagaInstanceState.AttachExistingEntity(loadedEntity);
             }
-
-            await next().ConfigureAwait(false);
```
Comment (andreasohlund) [inline src/NServiceBus.Core/Sagas/LoadSagaBehavior.cs]: Don't you still need to call next()?

Comment (danielmarbach) [inline src/NServiceBus.Core/Sagas/LoadSagaBehavior.cs]: Damn! Yes I do

**Comment by danielmarbach** [comment]
@andreasohlund fixed and rebased

**Comment by andreasohlund** [comment]
failing tests

**Comment by danielmarbach** [comment]
Will look into this some time

On Sunday, December 20, 2015, Andreas Öhlund <notifications@github.com>
wrote:

> failing tests
>
> On Sat, Dec 19, 2015 at 3:26 PM, Daniel Marbach <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>>
> wrote:
>
> > @andreasohlund <https://github.com/andreasohlund> fixed and rebased
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/Particular/NServiceBus/pull/3260#issuecomment-165988250
> >
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3260#issuecomment-166087062>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by danielmarbach** [comment]
Ok, realized what the problem is. `IHandleSagaNotFound` is flawed. As the interface and the doco suggests it is a general catch all solution when a saga is not found. I assumed that whenever a saga was not found all implementations of `IHandleSagaNotFound` should be invoked. BUT that is not the case as the two failing tests show.

Core assumes the following:

* When_sagas_cant_be_found.IHandleSagaNotFound_only_called_once

I'd say they should actually be called twice because `ReceiverWithSagas` has two sagas which could not be found

* When_sagas_cant_be_found.IHandleSagaNotFound_not_called_if_second_saga_is_executed

I'd say this is also wrong and should be called once since one saga was found and the other one not.

When it comes to executing sagas core treats sagas largely as message handlers (which is a faulty design as well but lets leave that aside). So in the above scenarios we have two sagas (message handlers) and each should be treated indiviudally when it comes to `IHandleSagaNotFound`. I think we have multiple ways to approach this:

* Rename `IHandleSagaNotFound` to `IHandleSagasNotFound` to better state the catch all nature of it
* Optionally: Introduce `IHandleSagaNotFound<T>` similar to `IFindSagas<T>` but withouth the generified message parameter to allow strongly typed not found handlers.

@andreasohlund thoughts? Should I raise a Short Term Focus for that?

**Comment by SzymonPobiega** [comment]
There is another serious flow when it comes to `IHandleSagaNotFound` which is, you can't reliably tell that something is not there. It was not there when we looked, but now it might actually be there. I think it might be time to actually implement the guidance that all messages need to be able to start a given saga. That would make the whole not found problem go away.

**Comment by andreasohlund** [comment]
Related to https://github.com/Particular/NServiceBus/issues/2787

**Comment by andreasohlund** [comment]
@danielmarbach we already have an issue, chime in there? https://github.com/Particular/PlatformDevelopment/issues/166

**Comment by andreasohlund** [comment]
>There is another serious flow when it comes to IHandleSagaNotFound which is, you can't reliably tell that something is not there. It was not there when we looked, but now it might actually be there

You bring up a good point @SzymonPobiega , since we're not writing anything to the persistence we don't get any concurrency checks which means that like you say the instance might just have been created but we have no way of detecting it

> I think it might be time to actually implement the guidance that all messages need to be able to start a given saga. That would make the whole not found problem go away.

:+1:

**Comment by SzymonPobiega** [comment]
Yeah. The only bullet proof alternative would be to actually insert a null
saga instance when invoking not found handlers. This way we would ensure
mutual exclusion of saga and not found handler execution.

**Comment by danielmarbach** [comment]
So then ditch this PR and address the larger issue?

**Comment by mikeminutillo** [comment]
This issue does not currently appear on the waffle board during syncs. Please find (or create) and issue on the [waffle board](https://waffle.io/particular/progress-strategic) for this and use the waffle ["Connect" keyword](https://github.com/waffleio/waffle.io/wiki/FAQs#prs-connect-keywords) to associate this issue with it.

**Pull request closed without merging**